### PR TITLE
Adds Groth16

### PR DESCRIPTION
**Pull request by kobigurk**

### CODE REVIEW

File: crypto-primitives/src/nizk/groth16/mod.rs [hunk 1]
```diff
@@ -0,0 +1,81 @@
...
+        end_timer!(proof_time);
+        Ok(result)
+    }
+
+    fn verify(
+        vk: &Self::PreparedVerificationParameters,
+        input: &Self::VerifierInput,
+        proof: &Self::Proof,
+    ) -> Result<bool, Error> {
+        let verify_time = start_timer!(|| "{Groth-Maller 2017}::Verify");
```
Comment (Pratyush) [inline crypto-primitives/src/nizk/groth16/mod.rs]:
```suggestion
        let verify_time = start_timer!(|| "{Groth 2016}::Verify");
```

File: crypto-primitives/src/nizk/groth16/mod.rs [hunk 2]
```diff
@@ -0,0 +1,81 @@
...
+        end_timer!(nizk_time);
+        Ok((pp, vk))
+    }
+
+    fn prove<R: Rng>(
+        pp: &Self::ProvingParameters,
+        input_and_witness: Self::AssignedCircuit,
+        rng: &mut R,
+    ) -> Result<Self::Proof, Error> {
+        let proof_time = start_timer!(|| "{Groth-Maller 2017}::Prove");
```
Comment (Pratyush) [inline crypto-primitives/src/nizk/groth16/mod.rs]:
```suggestion
        let proof_time = start_timer!(|| "{Groth 2016}::Prove");
```

File: crypto-primitives/src/nizk/groth16/mod.rs [hunk 3]
```diff
@@ -0,0 +1,81 @@
...
+        circuit: Self::Circuit,
+        rng: &mut R,
+    ) -> Result<
+        (
+            Self::ProvingParameters,
+            Self::PreparedVerificationParameters,
+        ),
+        Error,
+    > {
+        let nizk_time = start_timer!(|| "{Groth-Maller 2017}::Setup");
```
Comment (Pratyush) [inline crypto-primitives/src/nizk/groth16/mod.rs]:
```suggestion
        let nizk_time = start_timer!(|| "{Groth 2016}::Setup");
```

File: groth16/src/lib.rs [hunk 4]
```diff
@@ -0,0 +1,300 @@
+//! An implementation of the [Groth][Groth16] simulation extractable zkSNARK.
```
Comment (Pratyush) [inline groth16/src/lib.rs]:
```suggestion
//! An implementation of the [Groth][Groth16] zkSNARK.
```

File: groth16/src/generator.rs [hunk 5]
```diff
@@ -0,0 +1,355 @@
...
+
+    end_timer!(h_time);
+
+    // Compute the L-query
+    let l_time = start_timer!(|| "Calculate L");
+    let mut l_query = FixedBaseMSM::multi_scalar_mul::<E::G1Projective>(
+        scalar_bits,
+        g1_window,
+        &g1_table,
+    &l,
```
Comment (Pratyush) [inline groth16/src/generator.rs]:
```suggestion
        &l,
```

**Review by Pratyush — CHANGES_REQUESTED** [review]
Thanks for the PR, it looks great! Just a couple of additional comments:
- [x] Instead of adding the LICENSE files directly, could you instead make them symlinks like in the other crates?
- [x] Could you run the crate through `cargo fmt`? I noticed a couple of inconsistent indentation things.

**Comment by kobigurk** [comment]
All done!

I noticed that `cargo fmt` with the latest nightly did more changes than I expected. Do you have a particular config you're using?

If not, we should run the whole Zexe workspace through it.

**Comment by Pratyush** [comment]
I think it should use the rustfmt.toml that's in the repository root. From looking at the fmt commit, it seems that it used this config.

I usually run it on every crate except algebra because the formatting on constants gets screwed up by rustfmt.

**Comment by kobigurk** [comment]
Got it. I think we can leave the formatting to another PR so this wouldn't affect the whole repository?

**Comment by Pratyush** [comment]
Sure, as you wish!

**Comment by Pratyush** [comment]
Thanks for the PR @kobigurk! It's great to have another implementation of Groth16 in Rust =)

**Pull request merged**